thunder/distributed/transforms/fsdp_v2.py

```diff
@@ -153,92 +153,91 @@ def transform_module(
+        # Note that .named_parameters / .named_buffers used below only return a duplicated parameter once.
+
+        shared_names = thunder_model._get_shared_names()
+        self.shared_params_name = {}
+
+        # For materialized parameters and buffers, we move them to the target device as necessary
+        # for un-materialized parameters and buffers, we set the ._thunder_device
+
+        is_fully_materialized = True
+        for n, p in thunder_model.named_parameters():
+            for nn in shared_names[n]:
+                if nn != n:
+                    self.shared_params_name[nn] = n
+            try:
+                orig_p = thunder_model._model.get_parameter(n)
+            except AttributeError as _:
+                orig_p = None
+            if p.is_meta:
+                is_fully_materialized = False
+                p._thunder_device = self.device
+                if orig_p is not None:
+                    orig_p._thunder_device = self.device
+                # TODO: check if devic_adjustments are still needed
+                for nn in shared_names[n]:
+                    device_adjustments[nn] = self.device
+            elif p.device != self.device:
+                new_p = torch.nn.Parameter(p.to(device=self.device), requires_grad=p.requires_grad)
+                for nn in shared_names[n]:
+                    thunder_model._overrides_parameters[nn] = new_p
+                    device_adjustments[nn] = self.device
+
+        for n, b in thunder_model.named_buffers():
+            try:
+                orig_b = thunder_model._model.get_buffer(n)
+            except AttributeError as _:
+                orig_b = None
+            if b.is_meta:
+                is_fully_materialized = False
+                b._thunder_device = self.device
+                if orig_b is not None:
+                    orig_b._thunder_device = self.device
+                # TODO: check if this is still needed
+                device_adjustments[n] = self.device
+            elif b.device != self.device:
+                new_b = b.to(device=self.device)
+                for nn in shared_names[n]:
+                    thunder_model._overrides_buffers[nn] = new_p
+                    device_adjustments[nn] = self.device
+
+        # Broadcast parameters if requested
+        if self.broadcast_from is not None:
+            if not is_fully_materialized:
+                # Note: we could move broadcasting into its own transform coming
+                #       after materialization (in thunder.distributed.fsdp) to
+                #       support this, if it is useful.
+                raise RuntimeError("cannot broadcast from non-materialized model")
+            for pn, p in chain(thunder_model.named_parameters(), thunder_model.named_buffers()):
+                tdist.broadcast(
+                    p,
+                    src=self.broadcast_from,
+                    group=self.process_group,
+                    async_op=False,
```

Comment:
ditto, shouldn't this execution be under `torch.no_grad()`?

Reply:
Don't know. Was it before?

Reply:
added it, thank you!